rpc/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -101,31 +107,95 @@ impl OptimisticallyConfirmedBankTracker {
             optimistically_confirmed_bank,
             subscriptions,
             &mut pending_optimistically_confirmed_banks,
+            &mut last_notified_confirmed_slot,
+            &mut highest_confirmed_slot,
         );
         Ok(())
     }
 
+    fn notify_or_defer(
+        subscriptions: &Arc<RpcSubscriptions>,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        bank: &Arc<Bank>,
+        last_notified_confirmed_slot: &mut Slot,
+        pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+    ) {
+        if bank.is_frozen() {
+            if bank.slot() > *last_notified_confirmed_slot {
+                debug!(
+                    "notify_or_defer notifying via notify_gossip_subscribers for slot {:?}",
+                    bank.slot()
+                );
+                subscriptions.notify_gossip_subscribers(bank.slot());
+                *last_notified_confirmed_slot = bank.slot();
+            }
+        } else if bank.slot() > bank_forks.read().unwrap().root_bank().slot() {
+            pending_optimistically_confirmed_banks.insert(bank.slot());
+            debug!("notify_or_defer defer notifying for slot {:?}", bank.slot());
+        }
+    }
+
+    fn notify_or_defer_confirmed_banks(
+        subscriptions: &Arc<RpcSubscriptions>,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        bank: &Arc<Bank>,
+        slot_threshold: Slot,
+        mut last_notified_confirmed_slot: &mut Slot,
+        mut pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+    ) {
+        for parent in bank.parents().iter().rev() {
+            if parent.slot() > slot_threshold {
```

Comment:
```suggestion
        for parent in bank.parents().iter().take_while(|slot| slot > slot_threshold).rev() {
```
Just noticed, I think we can save a few iterations?

Reply:
Also, what do you think about using `Bank::parents_inclusive()` to avoid the dupe of `notify_or_defer` below?

Reply:
Thanks for the suggestion on take_while. But I am not convinced  take_while actually saves any real cycles -- looking at is implementation using TakeWhile --  it uses the wrapped iterator and apply the closure in the same way a normal iterator. Will look into using parents_inclusive.

Reply:
The difference is that `take_while` stops iterating when it reaches the first element that doesn't satisfy the predicate, whereas your code will continue on to check every parent in the list.
Here's a code snippet that demonstrates perf difference over a larger set of elements:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=73168b2c8b67710d0015cb8869a08bee

Reply:
Good point @CriesofCarrots -- will change that.

Reply:
Sadly, `take_while` doesn't support double-ended iterator; as-such, current code is more performative for the likely parents.len()